wiki/C++-Style-Guide.md

```diff
@@ -0,0 +1,814 @@
+
+## Avoid copy-and-pasted code
+
+If nearly identical code appears in two places, factor it out into a separate function.
+Otherwise, if you make a change or a bug fix, you're likely to make it in one of the copies but not in the other. And, copy-and-paste is often a sign of quickly written throw-away code.
+
+## Non-member functions
+
+All functions should either be a member of a class or of an unnamed namespace.
+This implies that some classes may consist entirely of static functions, and never actually be intended to be instantiated.
+This is so that it is immediately clear to anyone reading the code what file they need to look at to see the declaration or definition of that function. That is, if I see "MyClass::DoStuff(myData);", I know to look at MyClass.h for details. If I see "DoStuff(myData);", I have to read through all the #includes to see which one defined DoStuff().
+Functions in an unnamed namespace are only visible in a single .cpp file, so they will generally just be helper functions for that particular class.
+
+## Void
+
+Do not use "void" to denote a function with no arguments.
+An empty list of arguments is much clearer than a list of one element that happens to be void. The only reason people use "void" is as a carry-over from C, where these two function declarations mean different things:
+```c
+void F(void); // has zero arguments.
+void F(); // has an undefined number of arguments.
+```
+
+## Default arguments
+
+Use default arguments only when it is self-evident what the default ought to be.
+Using default arguments is preferable to defining several overloads of a function which all duplicate the same code.
+
+
+
+# Variables
+
+This section deals with naming of variables, scoping, initialization, etc.
+
+## Lower-case camelCase
+
+Variable names should be in camelCase, starting with a lower-case letter.
+This is so that variables can be easily distinguished from class and function names.
+
+## Constants in ALL_CAPS
+
+Constants may be in ALL_CAPITALS_WITH_UNDERSCORES.
+Or they can use the same format as normal variables.
+
+## Use const rather than #define
+
+Constants should be defined as "static const" rather than using macros.
+This avoids unintended side effects and is also more efficient if the variable in question has a non-trivial constructor, because this way one instance of it can be reused.
+
+## No Hungarian notation
+
+Do not use Hungarian notation (prefixes or suffixes on variable names just to represent their type).
+Hungarian notation makes variable names harder to read and harder to remember, because instead of English words they include random character sequences like "lptsz". Also, because the compiler is checking for type safety anyways, these prefixes do nothing to improve code quality.
```

Comment:
```suggestion
Hungarian notation makes variable names harder to read and harder to remember, because instead of English words they include random character sequences like `lptsz`. Also, because the compiler is checking for type safety anyways, these prefixes do nothing to improve code quality.
```